File: 프로그래머스/2/42626. 더 맵게/더 맵게.cpp
#include <iostream>
#include <queue>
#include <string>
#include <vector>

using namespace std;

int solution(vector<int> scoville, int K) {
    int cnt = 0, score = 0;
    priority_queue< int, vector<int>, greater<int> > pq(scoville.begin(), scoville.end());

    while (pq.top() < K)
    {
        if (pq.size() < 2)
            break;
        score = pq.top();
        pq.pop();
        score += pq.top() * 2;
        pq.pop();
        pq.push(score);
        cnt++;
    }
    if (pq.top() >= K)
        return cnt;
    else
        return -1;
}

File: 프로그래머스/2/42626. 더 맵게/더 맵게.java
import java.util.*;
    
class Solution {
    public int solution(int[] scoville, int K) {
        PriorityQueue<Long> pq = new PriorityQueue();
        int cnt = 0;
        
        for (int num : scoville) {
            pq.offer(Long.valueOf(num));
        }
        
        while (true) {
            Long first = pq.poll();
            
            if (first != null && first >= K)
                return cnt;
            
            Long second = pq.poll();
            
            if (first == null || second == null)
                return -1;
            else {
                pq.add(first + second * 2);
                cnt++;
            }
        }
    }
}